File: Lesson07/Notes/app/src/main/java/com/codingblocks/notes/MainActivity.java
package com.codingblocks.notes;

import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.Button;

import java.util.ArrayList;
import java.util.Calendar;

public class MainActivity extends AppCompatActivity {

    ArrayList<Note> noteArrayList = new ArrayList<>();
    NoteAdapter noteAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

//        Calendar c = Calendar.getInstance();
//
//        c.setTimeInMillis(System.currentTimeMillis());
//
//        c.getTime();

        RecyclerView recyclerView = findViewById(R.id.rvNotes);

        FloatingActionButton floatingActionButton = findViewById(R.id.fabAdd);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        noteAdapter = new NoteAdapter(noteArrayList);

        recyclerView.setAdapter(noteAdapter);

        floatingActionButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                Note note = new Note("Hello",
                        "World",
                        String.valueOf(System.currentTimeMillis()));

                noteArrayList.add(note);

//                noteAdapter.notifyDataSetChanged();

                noteAdapter.notifyItemInserted(noteArrayList.size() - 1);

            }
        });

    }
}

File: Lecture12/Coupling/app/src/main/java/com/codingblocks/roomdb/ui/activity/MainActivity.kt
package com.codingblocks.roomdb.ui.activity

import android.arch.persistence.room.Room
import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import android.view.LayoutInflater
import com.codingblocks.roomdb.model.Note
import com.codingblocks.roomdb.ui.adapter.ItemAdapter
import com.codingblocks.roomdb.R
import com.codingblocks.roomdb.`interface`.ClickHandlers
import com.codingblocks.roomdb.db.NoteDatabase
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.view_dialog.view.*

class MainActivity : AppCompatActivity(), ClickHandlers {

//    val db = Room.databaseBuilder(
//        this,
//        NoteDatabase::class.java,
//        "notesDb.db"
//    )
//        .build()

    private val db by lazy {
        Room.databaseBuilder(
            this,
            NoteDatabase::class.java,
            "notesDb.db"
        )
            .allowMainThreadQueries()
            .build()
    }

    private val notesList = arrayListOf<Note>()
    lateinit var itemAdapter: ItemAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        notesList.addAll(db.noteDao().getNotes())

        itemAdapter = ItemAdapter(notesList, this)

        rvNotes.layoutManager = LinearLayoutManager(this)

        rvNotes.adapter = itemAdapter

        floatingActionButton.setOnClickListener {
            showCustomAlert()
        }
    }

    private fun showCustomAlert() {

        val dialogView = LayoutInflater.from(this).inflate(R.layout.view_dialog, null, false)

        val customAlertDialog = AlertDialog.Builder(this)
            .setTitle("Hello from a custom alert")
            .setView(dialogView)
            .setPositiveButton("Ok") { dialog, which ->
                dialog.cancel()

                val title = dialogView.etTitle.text.toString()
                val subTitle = dialogView.etSubtitle.text.toString()

                val note = Note(title, subTitle)

                //Get the ID of the inserted note
                val id = db.noteDao().insertNote(note)  //if this ID is -1, that means the insertion failed

                //Update the note object so that it has an ID
                note.id = id.toInt()

                //Add the note with an ID to the arrayList
                notesList.add(0, note)

                rvNotes.smoothScrollToPosition(0)
                itemAdapter.notifyItemInserted(0)
            }
            .setNegativeButton("Close") { dialog, which ->
                dialog.cancel()
            }
            .create()

        customAlertDialog.show()
    }

    override fun onResume() {
        super.onResume()
        notesList.clear()
        notesList.addAll(db.noteDao().getNotes())
        itemAdapter.notifyDataSetChanged()
    }

    override fun handleClick(note: Note) {
        val intent = Intent(this, EditActivity::class.java)
        intent.putExtra("ID", note.id)
        Log.e("TAG", "ID is ${note.id}")
        startActivity(intent)
    }

    override fun handleLongClick(note: Note) {
        db.noteDao().deleteNote(note)
        notesList.remove(note)
        itemAdapter.notifyDataSetChanged()
    }

}
